plugins/CoreAdminHome/Commands/InvalidateReportData.php

```diff
@@ -0,0 +1,272 @@
+            self::ALL_OPTION_VALUE);
+        $this->addOption('cascade', null, InputOption::VALUE_NONE,
+            'If supplied, invalidation will cascade, invalidating child period types even if they aren\'t specified in'
+            . ' --periods. For example, if --periods=week, --cascade will cause the days within those weeks to be '
+            . 'invalidated as well. If --periods=month, then weeks and days will be invalidated. Note: if a period '
+            . 'falls partly outside of a date range, then --cascade will also invalidate data for child periods '
+            . 'outside the date range. For example, if --dates=2015-09-14,2015-09-15 & --periods=week, --cascade will'
+            . ' also invalidate all days within 2015-09-13,2015-09-19, even those outside the date range.');
+        $this->addOption('dry-run', null, InputOption::VALUE_NONE, 'For tests. Runs the command w/o actually '
+            . 'invalidating anything.');
+        $this->setHelp('Invalidate archived report data by date range, site and period. Invalidated archive data will '
+            . 'be re-archived during the next core:archive run. If your log data has changed for some reason, this '
+            . 'command can be used to make sure reports are generated using the new, changed log data.');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $invalidator = StaticContainer::get('Piwik\Archive\ArchiveInvalidator');
+
+        $cascade = $input->getOption('cascade');
+        $dryRun = $input->getOption('dry-run');
+
+        $sites = $this->getSitesToInvalidateFor($input);
+        $periodTypes = $this->getPeriodTypesToInvalidateFor($input);
+        $dateRanges = $this->getDateRangesToInvalidateFor($input);
+
+        $datesByPeriod = $this->getPeriodsToInvalidateFor($periodTypes, $dateRanges, $cascade);
+
+        foreach ($datesByPeriod as $periodType => $dates) {
+            $output->writeln("Invalidating $periodType periods...");
+
+            if ($output->getVerbosity() > OutputInterface::VERBOSITY_NORMAL) {
+                $output->writeln("  [ dates = " . implode(', ', $dates) . " ]");
+            }
+
+            if ($dryRun) {
+                $output->writeln("[Dry-run] invalidating archives for site = [ " . implode(', ', $sites)
+                    . " ], dates = [ " . implode(', ', $dates) . " ], period = [ $periodType ]");
+            } else {
+                $invalidationResult = $invalidator->markArchivesAsInvalidated($sites, $dates, $periodType);
+
+                if ($output->getVerbosity() > OutputInterface::VERBOSITY_NORMAL) {
+                    $output->writeln($invalidationResult->makeOutputLogs());
+                }
+            }
+        }
+    }
+
+    private function getSitesToInvalidateFor(InputInterface $input)
+    {
+        $sites = $input->getOption('sites');
+
+        $siteIds = Site::getIdSitesFromIdSitesString($sites);
```

Comment:
I think we have same code eg in `Access::checkUserHasAdminAccess` and possibly many other parts. Could be nice to extract it or maybe there is something that can be reused already